File: slide-rs/src/send.rs
use anyhow::{bail, Context, Result};
use console::style;
use indicatif::{ProgressBar, ProgressStyle};
use std::fs;
use std::io::Write;
use std::path::Path;
use std::thread;
use std::time::{Duration, Instant};

use crate::protocol::*;

pub fn send_session(port_name: &str, files: &[String], baud: u32, debug: bool) -> Result<()> {
    println!(
        "{} v0.2 — Serial Line Inter-Device Exchange",
        style("SLIDE").cyan().bold()
    );
    println!("  Port:  {} @ {} baud", style(port_name).yellow(), baud);
    println!("  Mode:  {}", style("send").green().bold());
    println!("  Files: {}", files.len());
    println!();

    // Validate files exist
    for f in files {
        if !Path::new(f).exists() {
            bail!("File not found: {f}");
        }
    }

    let mut port = open_serial(port_name, baud)?;

    // Handshake: sender sends RDY, waits for receiver's echo
    let spin = ProgressBar::new_spinner();
    spin.set_style(
        ProgressStyle::default_spinner()
            .template("{spinner:.cyan} {msg}")
            .unwrap(),
    );
    spin.set_message("Waiting for Z80 (start SLIDE R on Z80 now)...");
    spin.enable_steady_tick(Duration::from_millis(100));

    loop {
        port.write_all(&[CTRL_RDY])?;
        port.flush()?;
        thread::sleep(Duration::from_secs(1));
        match recv_control(port.as_mut(), Duration::from_secs(1)) {
            Ok(Control::Rdy) => break,
            _ => continue,
        }
    }
    spin.finish_with_message(format!("{} Z80 connected.", style("✓").green().bold()));
    thread::sleep(Duration::from_millis(50));
    port.clear(serialport::ClearBuffer::Input)?;

    // Send each file
    let session_start = Instant::now();
    let mut total_bytes = 0u64;

    for (i, filename) in files.iter().enumerate() {
        println!(
            "\n{}",
            style(format!(
                "── File {}/{}: {} ──",
                i + 1,
                files.len(),
                filename
            ))
            .dim()
        );
        let bytes = send_file(port.as_mut(), filename, debug)?;
        total_bytes += bytes;
    }

    // FIN exchange
    if debug {
        eprintln!("  DEBUG sending FIN");
    }
    port.write_all(&[CTRL_FIN])?;
    port.flush()?;
    let _ = recv_control(port.as_mut(), Duration::from_secs(5));

    let elapsed = session_start.elapsed();
    println!();
    println!(
        "{} {} file(s) sent, {} bytes in {:.1}s",
        style("✓").green().bold(),
        files.len(),
        total_bytes,
        elapsed.as_secs_f64()
    );

    Ok(())
}

fn send_file(port: &mut dyn serialport::SerialPort, filename: &str, debug: bool) -> Result<u64> {
    let file_data_raw = fs::read(filename).context("Reading file")?;
    let filesize = file_data_raw.len() as u32;

    // Pad to 128-byte boundary for CP/M
    let mut file_data = file_data_raw;
    if file_data.len() % 128 != 0 {
        file_data.resize(file_data.len() + (128 - file_data.len() % 128), 0x1A);
    }

    // Split into frames
    let mut frames: Vec<(u8, Vec<u8>)> = Vec::new();
    let mut offset = 0;
    let mut seq: u16 = 1; // seq 0 is the header
    while offset < file_data.len() {
        let end = (offset + FRAME_SIZE).min(file_data.len());
        frames.push(((seq & 0xFF) as u8, file_data[offset..end].to_vec()));
        seq += 1;
        offset += FRAME_SIZE;
    }

    let total_frames = frames.len();
    let display_name = Path::new(filename)
        .file_name()
        .unwrap_or_default()
        .to_string_lossy();
    println!(
        "  {} — {} bytes ({} frames)",
        style(&*display_name).bold(),
        filesize,
        total_frames
    );

    // Send header
    let header = build_header_frame(filename, filesize);
    if debug {
        eprintln!("  DEBUG header ({} bytes): {:02x?}", header.len(), &header);
    }
    port.write_all(&header)?;
    port.flush()?;

    let ctrl = recv_control(port, Duration::from_secs(60))?;
    match ctrl {
        Control::Can => bail!("Z80 rejected file — check disk/drive"),
        Control::Ack(_) => {}
        other => bail!("Expected ACK for header, got {other:?}"),
    }

    // Give Z80 time to create the file
    thread::sleep(Duration::from_millis(500));
    port.clear(serialport::ClearBuffer::Input)?;

    // Progress bar
    let pb = ProgressBar::new(total_frames as u64);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("  {bar:50.green/dim} {pos}/{len} frames  {msg}")
            .unwrap()
            .progress_chars("━╸─"),
    );

    // Stream data with sliding window
    let mut send_idx = 0;
    let eof_seq = if !frames.is_empty() {
        (frames.last().unwrap().0 as u16 + 1) as u8
    } else {
        1
    };
    let mut eof_sent = false;
    let start_time = Instant::now();

    while send_idx < total_frames {
        let window_end = (send_idx + WIN_SIZE).min(total_frames);

        for i in send_idx..window_end {
            let (seq_num, ref payload) = frames[i];
            let frame = build_frame(seq_num, payload);
            if debug {
                eprintln!(
                    "  DEBUG send frame seq={} len={} ({} bytes on wire)",
                    seq_num,
                    payload.len(),
                    frame.len()
                );
            }
            port.write_all(&frame)?;
        }

        // Append EOF to last window
        if window_end >= total_frames && !eof_sent {
            let eof_frame = build_frame(eof_seq, &[]);
            if debug {
                eprintln!("  DEBUG send EOF frame seq={eof_seq}");
            }
            port.write_all(&eof_frame)?;
            eof_sent = true;
        }
        port.flush()?;

        // Wait for response (skip RDY)
        let ctrl = loop {
            match recv_control(port, Duration::from_secs(10)) {
                Ok(Control::Rdy) => {
                    if debug {
                        eprintln!("  DEBUG got RDY (Z80 flushed to disk)");
                    }
                    continue;
                }
                Ok(c) => break c,
                Err(_) => {
                    pb.set_message(style("timeout, retrying...").yellow().to_string());
                    break Control::Rdy; // trigger retry
                }
            }
        };

        if debug {
            eprintln!("  DEBUG got {ctrl:?}");
        }

        match ctrl {
            Control::Can => {
                pb.abandon_with_message(style("disk error!").red().to_string());
                bail!("Z80 reported disk error — transfer aborted");
            }
            Control::Ack(acked_seq) => {
                // Advance send_idx past the acked sequence
                while send_idx < total_frames {
                    let fseq = frames[send_idx].0;
                    send_idx += 1;
                    if fseq == acked_seq {
                        break;
                    }
                }
                if acked_seq == eof_seq {
                    send_idx = total_frames;
                }
                pb.set_position(send_idx as u64);
            }
            Control::Nak(nak_seq) => {
                eof_sent = false;
                pb.set_message(
                    style(format!("NAK seq={nak_seq}, retransmitting..."))
                        .yellow()
                        .to_string(),
                );
                for (i, (fseq, _)) in frames.iter().enumerate() {
                    if *fseq == nak_seq {
                        send_idx = i;
                        break;
                    }
                }
            }
            _ => {} // RDY = retry the window
        }
    }

    // Send EOF if not already sent with last window
    if !eof_sent {
        let eof_frame = build_frame(eof_seq, &[]);
        port.write_all(&eof_frame)?;
        port.flush()?;
    }

    // Wait for final EOF ACK
    let _ = recv_control(port, Duration::from_secs(2));

    let elapsed = start_time.elapsed().as_secs_f64();
    let throughput = filesize as f64 / elapsed;
    let efficiency = throughput / (baud_to_bytes(port) as f64) * 100.0;

    pb.finish_with_message(format!(
        "{:.0} B/s ({:.0}% efficiency)",
        throughput, efficiency
    ));

    Ok(filesize as u64)
}

fn baud_to_bytes(port: &dyn serialport::SerialPort) -> u32 {
    port.baud_rate().unwrap_or(19200) / 10
}

File: slide-py/slide/send.py
#!/usr/bin/env python3
"""
SLIDE - Serial Line Inter-Device (file) Exchange
PC-side sender for custom Z80 serial transfer protocol.

Usage: slide-send <serial_port> <filename> [--baud 19200]

Protocol:
  Frame: [SOF=0x01] [SEQ] [LEN_H] [LEN_L] [PAYLOAD...] [CRC_H] [CRC_L]
  Control (Z80->PC): [ACK 0x06] [SEQ] | [NAK 0x15] [SEQ] | [RDY 0x11]
  Session: header frame -> streaming data frames -> zero-length end frame
"""

import sys
import os
import time
import argparse

from slide.common import (
    SOF, CTRL_ACK, CTRL_NAK, CTRL_RDY, CTRL_FIN, CTRL_CAN,
    WIN_SIZE, FRAME_SIZE,
    crc16_ccitt, build_frame, build_header_frame, recv_control, open_serial,
    _echo_can_and_drain,
)


def send_file(ser, filename: str, debug: bool = False):
    """Send a single file over an already-open serial connection."""

    filesize = os.path.getsize(filename)
    with open(filename, 'rb') as f:
        file_data = f.read()

    # Pad to 128-byte boundary for CP/M
    if len(file_data) % 128:
        file_data += b'\x1A' * (128 - (len(file_data) % 128))

    # Split into frames
    frames = []
    offset = 0
    seq = 1  # seq 0 is the header
    while offset < len(file_data):
        chunk = file_data[offset:offset + FRAME_SIZE]
        frames.append((seq & 0xFF, chunk))
        seq += 1
        offset += FRAME_SIZE

    total_frames = len(frames)
    print(f"  File: {filename}")
    print(f"  Size: {filesize} bytes ({total_frames} frames)")

    # --- Send header frame ---
    print("  Sending header frame...")
    header = build_header_frame(filename, filesize)
    if debug:
        print(f"    DEBUG header ({len(header)} bytes): {header.hex(' ')}")
    ser.write(header)
    ser.flush()

    ctrl, seq_ack = recv_control(ser, timeout=60.0)
    if debug:
        print(f"    DEBUG response: ctrl=0x{ctrl:02X} seq={seq_ack}")
    if ctrl == CTRL_CAN:
        print(f"  ERROR: Z80 rejected file — check disk/drive.")
        return False
    if ctrl != CTRL_ACK:
        print(f"  ERROR: Expected ACK for header, got {ctrl:#04x}")
        return False

    print("  Header acknowledged. Streaming data...\n")

    # Give Z80 time to create the file on disk before we start streaming
    time.sleep(0.5)
    ser.reset_input_buffer()

    # --- Stream data frames with sliding window ---
    send_idx = 0
    eof_seq = (frames[-1][0] + 1) & 0xFF if frames else 1
    eof_sent = False
    eof_acked = False
    start_time = time.time()

    while send_idx < total_frames:
        window_start = send_idx
        window_end = min(send_idx + WIN_SIZE, total_frames)

        for i in range(window_start, window_end):
            seq_num, payload = frames[i]
            frame = build_frame(seq_num, payload)
            if debug:
                print(f"    DEBUG sending frame seq={seq_num} len={len(payload)} ({len(frame)} bytes on wire)")
            ser.write(frame)

        # If this is the last window, append EOF so Z80 doesn't wait
        if window_end >= total_frames and not eof_sent:
            eof_frame = build_frame(eof_seq, b'')
            if debug:
                print(f"    DEBUG sending EOF frame seq={eof_seq}")
            ser.write(eof_frame)
            eof_sent = True

        ser.flush()

        # Wait for response (skip RDY signals, wait for ACK/NAK/CAN)
        retry = False
        while True:
            try:
                ctrl, resp_seq = recv_control(ser, timeout=10.0)
            except TimeoutError:
                print(f"\n  Timeout at frame {send_idx}. Retrying window...")
                retry = True
                break
            if ctrl == CTRL_CAN:
                print(f"\n\n  Z80 reported disk error — transfer aborted.")
                return False
            if ctrl != CTRL_RDY:
                break
            if debug:
                print(f"    DEBUG got RDY (Z80 flushed to disk)")
        if retry:
            continue

        if debug:
            ctrl_name = {CTRL_ACK: 'ACK', CTRL_NAK: 'NAK'}.get(ctrl, f'0x{ctrl:02X}')
            print(f"    DEBUG got {ctrl_name} seq={resp_seq}")

        if ctrl == CTRL_ACK:
            acked_seq = resp_seq
            while send_idx < total_frames:
                fseq = frames[send_idx][0]
                send_idx += 1
                if fseq == acked_seq:
                    break
            if acked_seq == eof_seq:
                eof_acked = True
                send_idx = total_frames

            pct = send_idx * 100 // total_frames
            bar = '#' * (pct // 2) + '-' * (50 - pct // 2)
            print(f"\r  [{bar}] {pct}% ({send_idx}/{total_frames})", end='')

        elif ctrl == CTRL_NAK:
            eof_sent = False
            print(f"\n  NAK received for seq {resp_seq}, retransmitting...")
            for i, (fseq, _) in enumerate(frames):
                if fseq == resp_seq:
                    send_idx = i
                    break

    # --- Send end-of-transfer (if not already sent with last window) ---
    if not eof_sent:
        eof_frame = build_frame(eof_seq, b'')
        ser.write(eof_frame)
        ser.flush()

    # Wait for the EOF ACK, properly.
    #
    # The receiver flushes its buffer to disk BEFORE acknowledging the EOF
    # frame, so this ACK can lag by seconds. It only lands here when the
    # frame count is an exact multiple of WIN_SIZE: the window ACK and the
    # EOF ACK are then two separate messages, the loop above exits on the
    # window ACK, and the EOF ACK arrives after it. Any other frame count
    # and the loop has already consumed it.
    #
    # This used to be a 2s courtesy wait whose result was discarded, which
    # meant FIN went out while the Z80 was still writing — the receiver then
    # answered the stale ACK sequence instead of echoing FIN.
    while not eof_acked:
        try:
            ctrl, ack_seq = recv_control(ser, timeout=30.0)
        except TimeoutError:
            print("\n  ERROR: no EOF ACK within 30s — the receiver may still "
                  "be writing to disk")
            return False
        if ctrl == CTRL_RDY:
            continue                     # keepalive during a disk flush
        if ctrl == CTRL_ACK:
            eof_acked = True
        elif ctrl == CTRL_NAK:
            print(f"\n  ERROR: receiver NAKed seq {ack_seq} after the EOF frame")
            return False
        elif ctrl == CTRL_CAN:
            print("\n  ERROR: receiver cancelled before acknowledging EOF")
            return False
        else:
            print(f"\n  ERROR: expected EOF ACK, got 0x{ctrl:02X}")
            return False

    elapsed = time.time() - start_time
    throughput = filesize / elapsed if elapsed > 0 else 0

    print(f"\n\n  Transfer complete!")
    print(f"  {filesize} bytes in {elapsed:.1f}s ({throughput:.0f} bytes/sec)")
    baud = ser.baudrate
    print(f"  Link efficiency: {throughput / (baud / 10) * 100:.0f}%")
    return True


def send_session(port: str, files: list, baud: int = 19200, debug: bool = False):
    """Send one or more files to the Z80 using SLIDE v0.2 session protocol."""

    print(f"SLIDE v0.2 - Serial Line Inter-Device Exchange (send)")
    print(f"  Port: {port} @ {baud} baud")
    print(f"  Files: {len(files)}")
    print()

    ser = open_serial(port, baud)

    # --- Handshake: sender sends RDY first, waits for receiver's RDY echo ---
    # Wakeup signature bytes (v0.2.1 §1) and other stray bytes are skipped.
    print("Waiting for Z80 (start SLIDE R on Z80 now)...")
    ser.timeout = 60
    while True:
        ser.write(bytes([CTRL_RDY]))
        ser.flush()
        time.sleep(1.0)
        # Check for an interesting reply
        ser.timeout = 1.0
        b = ser.read(1)
        if not b:
            continue
        if b[0] == CTRL_RDY:
            break
        if b[0] == CTRL_CAN:
            _echo_can_and_drain(ser)
            print("Cancelled by Z80 before handshake.")
            ser.close()
            return
    print("Z80 ready.")
    time.sleep(0.05)
    ser.reset_input_buffer()

    # --- Send each file ---
    for i, filename in enumerate(files):
        print(f"\n--- File {i+1}/{len(files)}: {filename} ---")
        ok = send_file(ser, filename, debug)
        if not ok:
            print("Session aborted.")
            ser.close()
            return

    # --- FIN exchange ---
    if debug:
        print("  DEBUG sending FIN")
    ser.write(bytes([CTRL_FIN]))
    ser.flush()
    # Wait for FIN echo
    try:
        ctrl, _ = recv_control(ser, timeout=5.0)
        if debug:
            print(f"  DEBUG got FIN echo: ctrl=0x{ctrl:02X}")
    except TimeoutError:
        pass

    print(f"\nSession complete — {len(files)} file(s) sent.")
    ser.close()


def main():
    parser = argparse.ArgumentParser(description='SLIDE - Send files to Z80/CP/M')
    parser.add_argument('port', help='Serial port (e.g., /dev/ttyUSB0, COM3)')
    parser.add_argument('files', nargs='+', help='File(s) to send')
    parser.add_argument('--baud', type=int, default=19200, help='Baud rate (default: 19200)')
    parser.add_argument('--debug', action='store_true', help='Show wire-level debug output')
    args = parser.parse_args()

    for f in args.files:
        if not os.path.exists(f):
            print(f"Error: file '{f}' not found")
            sys.exit(1)

    send_session(args.port, args.files, args.baud, args.debug)


if __name__ == '__main__':
    main()
